File: app/src/main/java/com/exercise/davismiyashiro/popularmovies/moviedetails/ReviewListAdapter.java
package com.exercise.davismiyashiro.popularmovies.moviedetails;

import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.exercise.davismiyashiro.popularmovies.R;
import com.exercise.davismiyashiro.popularmovies.data.Review;
import com.exercise.davismiyashiro.popularmovies.databinding.ReviewListItemBinding;

import java.util.List;

/**
 * Created by Davis Miyashiro on 01/03/2017.
 */

public class ReviewListAdapter extends RecyclerView.Adapter<ReviewListAdapter.ReviewHolder> {

    private List<Review> reviews;
    private OnReviewClickListener listener;

    public ReviewListAdapter(List<Review> reviews, OnReviewClickListener listener) {
        this.reviews = reviews;
        this.listener = listener;
    }

    public interface OnReviewClickListener {
        void onReviewClick(Review review);
    }

    @Override
    public ReviewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ReviewListItemBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.review_list_item, parent, false);
        return new ReviewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ReviewHolder holder, int position) {
        holder.binding.setReview(reviews.get(position));
        holder.binding.setReviewCallback(listener);
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return reviews.size();
    }

    public void replaceData(List<Review> reviews) {
        this.reviews = reviews;
        notifyDataSetChanged();
    }

    public class ReviewHolder extends RecyclerView.ViewHolder {

        ReviewListItemBinding binding;

        public ReviewHolder(ReviewListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/exercise/davismiyashiro/popularmovies/moviedetails/ReviewListAdapter.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Davis Miyashiro
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.exercise.davismiyashiro.popularmovies.moviedetails

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.exercise.davismiyashiro.popularmovies.data.Review
import com.exercise.davismiyashiro.popularmovies.databinding.ReviewListItemBinding

/**
 * Created by Davis Miyashiro on 01/03/2017.
 */

class ReviewListAdapter(
    private var reviews: List<Review>,
    private val listener: OnReviewClickListener
) :
    RecyclerView.Adapter<ReviewListAdapter.ReviewHolder>() {

    interface OnReviewClickListener {
        fun onReviewClick(review: Review)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReviewHolder {
        val binding =
            ReviewListItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ReviewHolder(binding)
    }

    override fun onBindViewHolder(holder: ReviewHolder, position: Int) {
        holder.bind(reviews[position])
    }

    override fun getItemCount(): Int {
        return reviews.size
    }

    fun replaceData(reviews: List<Review>?) {
        if (reviews != null) {
            this.reviews = reviews
            notifyDataSetChanged()
        }
    }

    inner class ReviewHolder(internal var binding: ReviewListItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(review: Review) {
            binding.reviewName.text = review.author
            binding.reviewContent.text = review.content
            binding.root.setOnClickListener { listener.onReviewClick(review) }
        }
    }
}
